File: python/basics/stack/155_min_stack.py
#https://leetcode.com/problems/min-stack/
'''
idea is to have another stack... which keeps the minimum..
if the next push is not the min.  push the latest.
pop... pop as normally
'''

'''
#its a modified version of stack copy.
    a tuple or sub-list is used to keep the count
'''
class MinStack_with_tuple:
    def __init__(self):
        self.stk = []
        self.min_stk = []

    def push(self, n):
        self.stk.append(n)
        #new minimum pushed..
        #if min. stack is empty or the new item is less than top of min stack
        if not self.min_stk or self.min_stk[len(self.min_stk)-1][0] > n:
            self.min_stk.append([n,1])
        elif n == self.min_stk[len(self.min_stk)-1][0]:
            #same item
            self.min_stk[len(self.min_stk)-1][1] += 1


    def pop(self):
        #reduce the count
        if self.min_stk[-1][0] == self.stk[-1]:
            self.min_stk[-1][1] -= 1
        if self.min_stk[-1][1] == 0:
            self.min_stk.pop()
        return self.stk.pop()


    def top(self):
        return self.stk[len(self.stk)-1]

    def get_min(self):
        return self.min_stk[len(self.min_stk)-1][0]




class MinStack:
    def __init__(self):
        self.stk = []
        self.min_stk = []

    def push(self, n):
        self.stk.append(n)
        if not self.min_stk or self.min_stk[len(self.min_stk)-1] > n:
            self.min_stk.append(n)
        else:
            self.min_stk.append(self.min_stk[len(self.min_stk)-1])


    def pop(self):
        self.min_stk.pop()
        return self.stk.pop()


    def top(self):
        return self.stk[len(self.stk)-1]

    def get_min(self):
        return self.min_stk[len(self.min_stk)-1]


#["MinStack","push","push","push","getMin","pop","top","getMin"]
#[[],[-2],[0],[-3],[],[],[],[]]
min_stack = MinStack()
min_stack.push(-2)
min_stack.push(0)
min_stack.push(-3)
min = min_stack.get_min()
print("min: " + str(min))

min_stack.pop()
top = min_stack.top()
print("top: " + str(top))

min = min_stack.get_min()
print("min: " + str(min))

#--------------------------------

min_stack = MinStack_with_tuple()
min_stack.push(-2)
min_stack.push(0)
min_stack.push(-3)
min = min_stack.get_min()
print("min: " + str(min))

min_stack.pop()
top = min_stack.top()
print("top: " + str(top))

min = min_stack.get_min()
print("min: " + str(min))

File: problem_solving/stack/cpp/155_min_stack.cpp
/*
#https://leetcode.com/problems/min-stack/
'''
idea is to have another stack... which keeps the minimum..
if the next push is not the min.  push the latest.
pop... pop as normally
'''

'''
#https://www.geeksforgeeks.org/design-a-stack-that-supports-getmin-in-o1-time-and-o1-extra-space/#approach
    push    2*x - min   ; min=x

    pop     min = 2*min - top

    peek    min > top, min, else top
'''

'''
#its a modified version of stack copy.
    a tuple or sub-list is used to keep the count
*/
//avoids storing the repeated values... by storing the count.
class MinStack_with_tuple{
    private:
    std::stack<int> main_stack;
    std::stack<int> min_stack;

    MinStack_with_tuple() {

    }   
    
    void push(int val) {

    }

    int pop() {

    }

    int top() {

    }

    int get_min() {

    }
};
    


class MinStack {
    private:
    std::stack<int> main_stack;
    std::stack<int> min_stack;

    public:
    MinStack() {

    }   
    
    void push(int val) {
        main_stack.push(val);
        if(min_stack.empty() || min_stack.top() >=  val) {
            min_stack.push(val);
        }
    }

    void pop() {
        if( !main_stack.empty() && !min_stack.empty() && min_stack.top() == main_stack.top() ) {
            min_stack.pop();
        }
        if(!main_stack.empty()) {
            main_stack.pop();
        }
    }

    int top() {
        if(!main_stack.empty()) {
            return main_stack.top();
        } 
        return -1;
    }

    int getMin() {
        if(!min_stack.empty()) {
            return min_stack.top();
        }
        return -1;
    }
};




//["MinStack","push","push","push","getMin","pop","top","getMin"]
//[[],[-2],[0],[-3],[],[],[],[]]
MinStack min_stack;
min_stack.push(-2);
min_stack.push(0);
min_stack.push(-3);
int min = min_stack.get_min();
std::cout << "min: " + min;

min_stack.pop();
top = min_stack.top();
std::cout << "top: " << top;

min = min_stack.get_min();
std::cout << "min: " + min << std::endl;

//--------------------------------

MinStack_with_tuple min_stack;
min_stack.push(-2);
min_stack.push(0);
min_stack.push(-3);
min = min_stack.get_min();
std::cout << "min: " + min << std::endl;

min_stack.pop();
top = min_stack.top();
std::cout << "top: " <<  top << std::endl;

min = min_stack.get_min();
std::cout << "min: " + min << std::endl;


